std/algorithm/iteration.d

```diff
@@ -4280,3 +4284,77 @@ unittest
          [1, 2, 0],
          [2, 1, 0]]));
 }
+
+/**
+Finds the mean (colloquially known as the average) of a range.
+Is $(D O(r.length)).
+
+Params:
+    r = a input range
+
+Returns:
+    the mean of r
+*/
+auto mean(R)(R r) if (isInputRange!R && !isInfinite!R) {
+    import std.typecons : tuple;
+
+    immutable auto pair = reduce!((a, b) => tuple(a[0] + 1, a[1] + b))(tuple(0u, 0.0), r);
+    return pair[1] / pair[0];
+}
+
+///
+@safe unittest {
+    assert(mean([1, 2, 3]) == 2);
+    assert(mean([1.5, 2.5, 12.5]) == 5.5);
+}
+
+@safe unittest {
+    import std.internal.test.dummyrange : ReferenceInputRange, ReferenceForwardRange;
+
+    auto r1 = new ReferenceInputRange!int([1, 2, 3]);
+    assert(r1.mean == 2);
+
+    auto r2 = new ReferenceInputRange!double([1.5, 2.5, 12.5]);
+    assert(r2.mean == 5.5);
+}
+
+/**
+Finds the median of a random access range. Is $(D Ο(r.length * log(r.length))).
+
+Params:
+    r = a random access range
+
+Returns:
+    the median of r
+*/
+auto median(R)(R r)
+if (isRandomAccessRange!R &&
+    hasSwappableElements!R &&
+    !isInfinite!R)
+{
+    import std.algorithm.sorting : sort;
+
+    if (r.length == 1)
+    {
+        return r[0];
+    }
+
+    auto sorted = sort(r.save);
```

Comment:
Would it be better to make a copy of the data and then sort it, or would it be better to just have the function mutate the range?